pkg/kv/kvserver/client_replica_circuit_breaker_test.go

```diff
@@ -0,0 +1,566 @@
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package kvserver_test
+
+import (
+	"context"
+	"sync/atomic"
+	"testing"
+	"time"
+
+	"github.com/cockroachdb/cockroach/pkg/base"
+	"github.com/cockroachdb/cockroach/pkg/keys"
+	"github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord"
+	"github.com/cockroachdb/cockroach/pkg/kv/kvserver"
+	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/kvserverbase"
+	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/liveness"
+	"github.com/cockroachdb/cockroach/pkg/roachpb"
+	"github.com/cockroachdb/cockroach/pkg/server"
+	"github.com/cockroachdb/cockroach/pkg/testutils"
+	"github.com/cockroachdb/cockroach/pkg/testutils/skip"
+	"github.com/cockroachdb/cockroach/pkg/testutils/testcluster"
+	"github.com/cockroachdb/cockroach/pkg/util/circuit"
+	"github.com/cockroachdb/cockroach/pkg/util/hlc"
+	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/cockroach/pkg/util/stop"
+	"github.com/cockroachdb/errors"
+	"github.com/stretchr/testify/require"
+)
+
+func TestReplicaCircuitBreaker(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	ctx := context.Background()
+	tc := setupCircuitBreakerTest(t)
+	defer tc.Stopper().Stop(ctx)
+
+	// In all scenarios below, we are starting out with our range on n1 and n2,
+	// and all other ranges (in particular the liveness range) on n1.
+	const (
+		n1 = 0
+		n2 = 1
+
+		pauseHeartbeats = true
+		keepHeartbeats  = true
+	)
+
+	// This is a sanity check in which the breaker plays no role.
+	runCircuitBreakerTest(t, "breaker-ok", func(t *testing.T, ctx context.Context, tc *circuitBreakerTest) {
+		// Circuit breaker doesn't get in the way of anything unless
+		// something trips it.
+		require.NoError(t, tc.Write(n1))
+		tc.RequireIsNotLeaseholderError(t, tc.Write(n2))
+		require.NoError(t, tc.Read(n1))
+		tc.RequireIsNotLeaseholderError(t, tc.Read(n2))
+	})
+
+	// In this test, n1 holds the lease and we disable the probe and trip the
+	// breaker. While the breaker is tripped, requests fail-fast with either a
+	// breaker or lease error. When the probe is re-enabled, everything heals.
+	runCircuitBreakerTest(t, "leaseholder-tripped", func(t *testing.T, ctx context.Context, tc *circuitBreakerTest) {
+		// Get lease on n1.
+		require.NoError(t, tc.Write(n1))
+		// Disable the probe so that when the breaker trips, it stays tripped.
+		tc.SetProbeEnabled(n1, false)
+		tc.Report(n1, errors.New("boom"))
+
+		// n1 could theoretically still serve reads (there is a valid lease
+		// and none of the latches are taken), but since it is hard to determine
+		// that upfront we currently fail all reads as well.
+		tc.RequireIsBreakerOpen(t, tc.Read(n1))
+		tc.RequireIsBreakerOpen(t, tc.Write(n1))
```

Comment:
I wonder if we'd want to add a latency assertion for these operations, to make sure we actually fail fast. These can be pretty annoying to test reliably, but it's an important property of the breaker that requests don't appear to hang.